src/tokio/seqpacket.rs

```diff
@@ -0,0 +1,200 @@
+use crate::nonblocking;
+use crate::UnixSocketAddr;
+use futures::ready;
+use std::io;
+use std::mem::MaybeUninit;
+use std::path::Path;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio_02::future::poll_fn;
+use tokio_02::io::{AsyncRead, AsyncWrite, PollEvented};
+
+/// An I/O object representing a Unix Sequenced-packet socket.
+pub struct UnixSeqpacketConn {
+    io: PollEvented<nonblocking::UnixSeqpacketConn>,
+}
+
+impl UnixSeqpacketConn {
+    /// Connects to the socket named by path.
+    ///
+    /// This function will create a new Unix socket and connect to the path
+    /// specified, associating the returned stream with the default event loop's
+    /// handle.
+    pub async fn connect<P: AsRef<Path>>(path: P) -> io::Result<UnixSeqpacketConn> {
+        let conn = nonblocking::UnixSeqpacketConn::connect(path)?;
+        let conn = UnixSeqpacketConn::new(conn)?;
+
+        //poll_fn(|cx| conn.io.poll_write_ready(cx)).await?;
+        Ok(conn)
+    }
+
+    pub(crate) fn new(conn: nonblocking::UnixSeqpacketConn) -> io::Result<UnixSeqpacketConn> {
+        let io = PollEvented::new(conn)?;
+        Ok(UnixSeqpacketConn { io })
+    }
+}
+
+impl UnixSeqpacketConn {
+    /// Sends data on the socket to the socket's peer.
+    pub async fn send(&mut self, buf: &[u8]) -> io::Result<usize> {
+        poll_fn(|cx| self.poll_send_priv(cx, buf)).await
+    }
+
+    /// Receives data from the socket.
+    pub async fn recv(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        poll_fn(|cx| self.poll_recv_priv(cx, buf)).await
+    }
+
+    pub(crate) fn poll_recv_priv(
+        &self,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        ready!(self.io.poll_read_ready(cx, mio::Ready::readable()))?;
+
+        match self.io.get_ref().recv(buf) {
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                self.io.clear_read_ready(cx, mio::Ready::readable())?;
+                Poll::Pending
+            }
+            Err(e) => Poll::Ready(Err(e)),
+            Ok((x, _truncated)) => Poll::Ready(Ok(x)),
+        }
+    }
+
+    pub(crate) fn poll_send_priv(
+        &self,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        ready!(self.io.poll_write_ready(cx))?;
+
+        match self.io.get_ref().send(buf) {
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                self.io.clear_write_ready(cx)?;
+                Poll::Pending
+            }
+            x => Poll::Ready(x),
+        }
+    }
+}
+
+impl AsyncRead for UnixSeqpacketConn {
+    unsafe fn prepare_uninitialized_buffer(&self, _: &mut [MaybeUninit<u8>]) -> bool {
+        false
+    }
+
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        self.poll_recv_priv(cx, buf)
+    }
+}
+
+impl AsyncWrite for UnixSeqpacketConn {
```

Comment:
I touch upon the reasoning for this in the [notes](https://github.com/tormol/uds/pull/2/files#diff-145eba58ca323c6c0ec060ca120c3ebbR3-R9), but it is definitely a point of discussion, and some contention especially surrounding the codec/`Framed` API and resulting functionality like `Sink` and `Stream`.

It seems that tokio's `UnixDatagram` implementation (which serves as the basis for this comparison because of its similarity to `Seqpacket`) has over time gotten less attention than its `UnixStream` counterpart. My coworker actually [introduced `UnixDatagramFramed`](https://github.com/tokio-rs/tokio/pull/453) since, similar to your point, `AsyncRead`/`Write` doesn't really make sense in the context of datagrams, so a completely separate API was required for codec support. But as far as I can tell, the tokio 0.2 migration did not carry over this functionality, perhaps due to lack of interest, and I suspect that any implementation could be subject to further breakage as tokio 0.3, 1.0, etc. are forthcoming.

This leads the discussion to `Seqpacket`-- I took a bit of a hand-wavy shortcut by saying that since sequenced packets behave mostly like unix streams, then the same API could be applied with the understanding, at the application layer, that sender and receiver should have some mutual understanding of the packet boundaries. This allows for tapping into useful, composable features like codec, `Stream[Ext]`, etc. for "free" since everything is already built for stream-type sockets. That said, I'm happy to go the `Datagram` route of having `UnixSeqpacketConnFramed`, and then impl'ing `Sink` and `Stream` on top of that. There does at least exist a [`UdpFramed`](https://docs.rs/tokio-util/0.3.1/tokio_util/udp/struct.UdpFramed.html) type whose own decoupled API could be modeled after.

Reply:
I brought up this potential solution in the tokio discord, and the feedback wasn't very positive. And I also learned that `UdpFramed` going away, so I think I'll just drop these trait impl's, and punt on any kind of codec support in this PR.